tests/operator_test.go

```diff
@@ -65,3 +69,103 @@ func TestCRDExplainable(t *testing.T) {
 	Expect(out).To(ContainSubstring("OperatorVersion The version of the HostPathProvisioner Operator"))
 	Expect(out).To(ContainSubstring("TargetVersion The targeted version of the HostPathProvisioner deployment"))
 }
+
+func TestNodeSelector(t *testing.T) {
+	RegisterTestingT(t)
+	tearDown, k8sClient := setupTestCase(t)
+	defer tearDown(t)
+
+	hppClient, err := getHPPClient()
+	Expect(err).ToNot(HaveOccurred())
+
+	cr, err := hppClient.HostpathprovisionerV1beta1().HostPathProvisioners().Get(context.TODO(), "hostpath-provisioner", metav1.GetOptions{})
+	Expect(err).ToNot(HaveOccurred())
+
+	nodes, _ := k8sClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	affinityTestValue := &v1.Affinity{
+		NodeAffinity: &v1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+				NodeSelectorTerms: []v1.NodeSelectorTerm{
+					{
+						MatchExpressions: []v1.NodeSelectorRequirement{
+							{Key: "kubernetes.io/hostname", Operator: v1.NodeSelectorOpIn, Values: []string{nodes.Items[0].Name}},
+						},
+					},
+				},
+			},
+		},
+	}
+	nodeSelectorTestValue := map[string]string{"kubernetes.io/arch": "not-a-real-architecture"}
+	tolerationsTestValue := []v1.Toleration{{Key: "test", Value: "123"}}
+
+	origWorkloads := cr.Spec.Workloads.DeepCopy()
+	defer func() {
+		cr, err = hppClient.HostpathprovisionerV1beta1().HostPathProvisioners().Get(context.TODO(), "hostpath-provisioner", metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		cr.Spec.Workloads = *origWorkloads.DeepCopy()
+
+		_, err = hppClient.HostpathprovisionerV1beta1().HostPathProvisioners().Update(context.TODO(), cr, metav1.UpdateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		err = k8sClient.AppsV1().DaemonSets("hostpath-provisioner").Delete(context.TODO(), "hostpath-provisioner", metav1.DeleteOptions{})
```

Comment:
This delete should not be necessary unless there is a bug in the operator that doesn't fix the daemonset during reconciliation.

Reply:
it's probably because we merge the updated daemonset and existing one.
you could argue that that's a different way to configure things, but it's kinda stateful and more likely to be lost.

With https://gist.github.com/maya-r/2e82d9b1a1955c27a2cfb32882e1b41d I don't need to delete the daemonset, but there are some extra fields in the spec filled out by kubernetes that aren't in our desired DaemonSet so it tries to update the DaemonSet on every loop, which isn't ideal.

I could add them, but what if a new one is added in a kubernetes update? seems non-ideal.

Reply:
OK, I have a pending change in https://github.com/kubevirt/hostpath-provisioner-operator/pull/94
I discovered "DeepDerivative" as opposed to "DeepEqual" along the way but it still didn't let me restore it to an empty value. I'm just explicitly setting the values after merge now.